File: Placement Sheet/LinkedList/BS/SortLinkedList012.ts
class NodeLL {
    private data: number;
    private next: NodeLL | null;
    constructor(data: number) {
        this.data = data;
        this.next = null;
    }

    getData(): number {
        return this.data;
    }

    getNext(): NodeLL | null {
        return this.next;
    }

    setNext(node: NodeLL | null): void {
        this.next = node;
    }
}

class SortLinkedList012 {
    private head: NodeLL | null = null;

    //Insert...
    insert(data: number): void {
        let newNode: NodeLL | null = new NodeLL(data);
        if (this.head == null) {
            this.head = newNode;
            return;
        }
        let currNode: NodeLL | null = this.head;
        while (currNode.getNext() !== null) {
            currNode = currNode?.getNext()!;
        }
        currNode.setNext(newNode);
    }

    //Display...
    display(): void {
        if (this.head == null) {
            console.log("List is empty...");
            return;
        }
        let currNode: NodeLL | null = this.head;
        let result: string = "";
        while (currNode !== null) {
            result += currNode.getData() + " --> ";
            currNode = currNode.getNext();
        }
        result += "Null";
        console.log(result);
    }

    //Sort012...
    sort012(): void {
        if (this.head == null) {
            console.log("List is empty...");
            return;
        }
        let zeroD: NodeLL | null = new NodeLL(0), oneD: NodeLL | null = new NodeLL(0), twoD: NodeLL | null = new NodeLL(0);
        let zero = zeroD, one = oneD, two = twoD;

        let currNode: NodeLL | null = this.head;
        while (currNode !== null) {
            if (currNode.getData() === 0) {
                zero.setNext(currNode);
                zero = zero.getNext()!;
            } else if (currNode.getData() === 1) {
                one.setNext(currNode);
                one = one.getNext()!;
            } else {
                two.setNext(currNode);
                two = two.getNext()!;
            }
            currNode = currNode.getNext();
        }
        zero.setNext(oneD.getNext() ? oneD.getNext() : twoD.getNext());
        one.setNext(twoD.getNext());
        two.setNext(null);
        this.head = zeroD.getNext();
    }
}

let list: SortLinkedList012 = new SortLinkedList012();
list.insert(1);
list.insert(2);
list.insert(0);
list.insert(1);
list.insert(2);
list.insert(0);
console.log("Before sorting: ");
list.display();
console.log("After sorting: ");
list.sort012();
list.display();

export { };

File: Placement Sheet/LinkedList/BS/SortLinkedList012.java
public class SortLinkedList012 {
    public static class Node {
        int data;
        Node next;

        Node(int data) {
            this.data = data;
            this.next = null;
        }
    }

    private Node head = null;
    private Node tail = null;

    // Insert...
    public void insert(int data) {
        Node newNode = new Node(data);
        if (this.head == null) {
            this.head = newNode;
            this.tail = newNode;
            return;
        }
        tail.next = newNode;
        tail = newNode;
    }

    // Display...
    public void display() {
        if (head == null) {
            System.out.println("List is empty...");
            return;
        }
        Node currNode = head;
        while (currNode != null) {
            System.out.print(currNode.data + " --> ");
            currNode = currNode.next;
        }
        System.out.println("Null");
    }

    public void sort012() {
        if (head == null) {
            System.out.println("List is empty...");
            return;
        }
        Node zeroD = new Node(0), oneD = new Node(0), twoD = new Node(0);
        Node zero = zeroD, one = oneD, two = twoD;
        Node currNode = head;
        while (currNode != null) {
            if (currNode.data == 0) {
                zero.next = currNode;
                zero = zero.next;
            } else if (currNode.data == 1) {
                one.next = currNode;
                one = one.next;
            } else {
                two.next = currNode;
                two = two.next;
            }
            currNode = currNode.next;
        }
        zero.next = oneD.next != null ? oneD.next : twoD.next;
        one.next = twoD.next;
        two.next = null;
        head = zeroD.next;
    }

    public static void main(String[] args) {
        SortLinkedList012 list = new SortLinkedList012();
        list.insert(1);
        list.insert(2);
        list.insert(0);
        list.insert(1);
        list.insert(2);
        list.insert(0);
        System.out.print("Before sorting: ");
        list.display();
        System.out.print("After sorting: ");
        list.sort012();
        list.display();
    }
}